categories/category.lean

```diff
@@ -0,0 +1,26 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Stephen Morgan, Scott Morrison
+
+namespace categories
+
+structure {u v} Category :=
+  ( Obj : Type u )
+  ( Hom : Obj → Obj → Type v )
+  ( identity : Π X : Obj, Hom X X )
+  ( compose  : Π { X Y Z : Obj }, Hom X Y → Hom Y Z → Hom X Z )
+
+  ( left_identity  : ∀ { X Y : Obj } (f : Hom X Y), compose (identity X) f = f )
+  ( right_identity : ∀ { X Y : Obj } (f : Hom X Y), compose f (identity Y) = f )
+  ( associativity  : ∀ { W X Y Z : Obj } (f : Hom W X) (g : Hom X Y) (h : Hom Y Z),
+    compose (compose f g) h = compose f (compose g h) )
```

Comment:
Stylistic suggestions: 
* Don't indent the structure fields, and don't put spaces around `(x : T)` and `{x : T}`.
* Don't use uppercase letters in identifiers - for the most part lean naming convention is all `snake_case`. The exception is "bundled" structures, and admittedly `Category` is one.

But I would like to investigate having `Obj` be a parameter instead of a member of `Category`, so you get `category A`. Then it can be a typeclass. Also, can we have the Hom set live in the same universe? This will cause universe inference issues as it is.

Reply:
> But I would like to investigate having `Obj` be a parameter instead of a member of `Category`, so you get `category A`. Then it can be a typeclass.

The style that I quite like is to have the category be uniquely defined by the homomorphism, from which the object type is inferred:

```lean
class category (H : O → O → Type v) extends has_comp H :=
  (id       : ∀ {a : O}, H a a)
  (left_id  : ∀ {a b : O} (f : H a b), id ∘ f = f)
  (right_id : ∀ {a b : O} (f : H a b), f ∘ id = f)
  (assoc    : ∀ {a b c d : O} (h : H c d) (g : H b c) (f : H a b), h ∘ (g ∘ f) = (h ∘ g) ∘ f)
```

I chose the field names to be in line with [`lean/library/init/function.lean`](https://github.com/leanprover/lean/blob/0fe561d063da40b862b181bd92ec9685133e33bd/library/init/function.lean).

I like the `∘` notation, so I put composition into a separate type class, which may or may not be useful:

```lean
class has_comp (H : O → O → Type v) :=
  (comp : ∀ {a b c : O}, H b c → H a b → H a c)

instance has_comp.arrow : has_comp (→) :=
  { comp := @function.comp
  }

-- Convenient composition notation.
infixr ` ∘ ` := has_comp.comp _
```

I adapted my [`category.lean`](https://github.com/spl/alpha-conversion-is-easy/blob/1785c9fe9d6a76f82a8bfa2d68370b9b6bb52979/src/data/category.lean) from @cipher1024's [`category.lean`](https://github.com/unitb/lean-lib/blob/b4334d79a829bc93d222250bfe38d83ab9f5d2e6/src/util/category.lean).

Reply:
The problem with this is that in many cases the type `O` *can't* be inferred from `H`, for example if `H` is `(<)` then it's still no clearer what the base type is. The arrow is probably a special case here since the base type is `Sort*` (and lean figures out the universes by a different method).

Reply:
The basic difficulty with having `Obj` and `Hom` live in the same universe, is that it’s then hard to define the category of types. If we have `Obj := Type u` and `Hom := ->`, then `Obj` lives in Type u+1 while Hom is still in u. Perhaps we could use ulift to raise Hom up one level to match.

I’ve run into quite a few other situations where natural operations on categories result in Obj and Hom living at different universes, even when they started the same. These could all be patched up using `ulift`, but I’m not certain it’s a good idea.